keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -107,123 +116,81 @@ serves the resourceVersion="0" list requests from reflectors today.
-it minimizes the chance that these clients will end up with an out-of-date
-resource version and need to relist (which can impact scalability). See [Watch
-Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
-for details.
+As a consequence of this number of etcd progress notification will increase
+number. [Watch Bookmarks KEP] utilized etcd periodic progress notifications to
+notify kube-apiserver clients of their progress. It used periodic watch
+notifications that are naturally limited on etcd side. To maintain the same
+behavior we will limit watch bookmarks to send them only one per 5 seconds.
 
-#### Determining if etcd is sending progress notify events
-
-It is possible to automatically determine if etcd is sending progress notify events.
-Each watch cache could keep track of when it received the last progress notify event.
-If it has been sufficiently long since the last one was received, or if none have
-ever been received, the watch cache should assume etcd is not sending progress notify,
-and not attempt to serve consistent reads from cache, falling back to serving
-the reads directly from etcd.
+[getCurrentResourceVersionFromStorage]: https://github.com/kubernetes/kubernetes/blob/3f247e59edfd4083242ad7271d076a38291760ff/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1246-L1278
+[Watch-List KEP]: /keps/sig-api-machinery/3157-watch-list
+[Watch Bookmarks KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/956-watch-bookmark
 
 ### Risks and Mitigations
 
-Configuring etcd to send progress notify events may have performance implications.
-@mm4tt and @wojtek-t have run [some
-experiments](https://github.com/kubernetes/kubernetes/pull/86769) that suggest
-that a progress notify interval of 50ms results in a noticeable increase in both
-etcd and kube-apiserver CPU utilization and some increased kube-apiserver serving
-latency. We intend to address this by first trying a 250ms progress notify interval,
-if ths performs well we will use that as our interval. If not, we will dig in
-to figure out why the events are impacting performance and see if we can optimize
-it away.
+### Dependency on manual watch progress notifications
 
-## Design Details
+<<[UNRESOLVED @serathius]>>
+Validate performance implications of using `WatchProgressRequest`.
+<<[/UNRESOLVED]>>
 
-### Progress notify interval selection
+<<[UNRESOLVED @serathius]>>
+Propose how kube-apiserver should fallback if etcd watch notification doesn't work:
+* Flag providing etcd version to kube-apiserver `--storage-backend=etcd3.4`
+* Flag to enable/disable the feature `--allow-using-progress-notify`
+* Have kube-apiserver check cluster version in etcd `/version` endpoint.
+  Retry the check logic if `WatchProgressRequest` fails.
+* Fallback to reading from etcd if no progress notification within `X` seconds.
+  <<[/UNRESOLVED]>>
```

Comment:
Please add a comment that I also requested above:
```
v3.5.8 and v3.4.25 that are 2 months old releases.
````

So etcd minor version isn't enough, because e.g. 3.4.21 has a bug

Reply:
Added a information about a bug.